src/ripple/app/ledger/Ledger.h

```diff
@@ -125,106 +89,217 @@ class Ledger
+    */
+    void
+    insert (SLE const& sle);
+
+    /** Fetch a modifiable state SLE.
+        Effects:
+            Gives the caller ownership of an
+                unflattened copy of the SLE.
+        @param type An optional LedgerEntryType. If type is
+                    engaged and the SLE's type does not match,
+                    then boost::none is returned.
+        @return `empty` if the key is not present
+    */
+    boost::optional<SLE>
+    fetch (uint256 const& key, boost::optional<
+        LedgerEntryType> type = boost::none) const;
+
+    // DEPRECATED
+    // Retrieve immutable ledger entry
+    SLE::pointer getSLEi (uint256 const& uHash) const;
+
+    /** Replace an existing state SLE.
+        Effects:
+            assert if key does not already exist.
+            The previous flattened SLE associated with
+                the key is released.
+            The key in the state map is associated
+                with a flattened copy of the SLE.
+        @note The key is taken from the SLE
+    */
+    void
+    replace (SLE const& sle);
+
+    /** Remove an state SLE.
+        Effects:
+            assert if the key does not exist.
+            The flattened SLE associated with the key
+                is released from the state map.
+    */
+    void
+    erase (uint256 const& key);
+
+    //--------------------------------------------------------------------------
+
+    void
+    updateHash();
 
-    void updateHash ();
     void setClosed ()
     {
         mClosed = true;
     }
+    
     void setValidated()
     {
         mValidated = true;
     }
-    void setAccepted (
-        std::uint32_t closeTime, int closeResolution, bool correctCloseTime);
+    
+    void setAccepted (std::uint32_t closeTime,
+        int closeResolution, bool correctCloseTime);
 
     void setAccepted ();
+
     void setImmutable ();
+
     bool isClosed () const
     {
         return mClosed;
     }
+
     bool isAccepted () const
     {
         return mAccepted;
     }
+
     bool isValidated () const
     {
         return mValidated;
     }
+
     bool isImmutable () const
     {
         return mImmutable;
     }
-    bool isFixed () const
-    {
-        return mClosed || mImmutable;
-    }
+
     void setFull ()
     {
-        mTransactionMap->setLedgerSeq (mLedgerSeq);
-        mAccountStateMap->setLedgerSeq (mLedgerSeq);
+        mTransactionMap->setLedgerSeq (seq_);
+        mAccountStateMap->setLedgerSeq (seq_);
     }
 
     // ledger signature operations
     void addRaw (Serializer& s) const;
     void setRaw (SerialIter& sit, bool hasPrefix);
 
-    uint256 const& getHash ();
+    /** Return the hash of the ledger.
+        This will recalculate the hash if necessary.
+    */
+    uint256 const&
+    hash();
+
+    // DEPRECATED
+    uint256 const&
+    getHash()
+    {
+        return hash();
+    }
+
+    /** Return the hash of the ledger. */
+    uint256 const&
+    hash() const;
```

Comment:
I don't like that `hash` can give different values depending on if the variable is const or not. i.e.
```a.hash()``` will recompute `mTransHash` and `mAccountHash` if non-const and will use the current values if const.